alibi_detect/utils/pytorch/kernels.py

```diff
@@ -88,6 +168,271 @@ def forward(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch
+        sigma
+            Bandwidth used for the kernel.
+        """
+        super().__init__()
+        self.parameter_dict['tau'] = 'period'
+        self.parameter_dict['sigma'] = 'bandwidth'
+        if tau is None:
+            self.log_tau = nn.Parameter(torch.empty(1), requires_grad=trainable)
+            self.init_required = True
+        else:
+            self.log_tau = nn.Parameter(tau.log(), requires_grad=trainable)
+            self.init_required = False
+        if sigma is None:
+            self.log_sigma = nn.Parameter(torch.empty(1), requires_grad=trainable)
+            self.init_required = True
+        else:
+            self.log_sigma = nn.Parameter(sigma.log(), requires_grad=trainable)
+            self.init_required = False
+        self.init_fn_tau = init_fn_tau
+        self.init_fn_sigma = init_fn_sigma
+        if active_dims is not None:
+            self.active_dims = torch.tensor(active_dims)
+        else:
+            self.active_dims = None
+        self.feature_axis = feature_axis
+        self.trainable = trainable
+
+    @property
+    def tau(self) -> torch.Tensor:
+        return self.log_tau.exp()
+
+    @property
+    def sigma(self) -> torch.Tensor:
+        return self.log_sigma.exp()
+
+    def forward(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
+                infer_parameter: bool = False) -> torch.Tensor:
+        x, y = torch.as_tensor(x), torch.as_tensor(y)
+        if self.active_dims is not None:
+            x = torch.index_select(x, self.feature_axis, self.active_dims)
+            y = torch.index_select(y, self.feature_axis, self.active_dims)
+        dist = torch.sqrt(distance.squared_pairwise_distance(x.flatten(1), y.flatten(1)))
+
+        if infer_parameter or self.init_required:
+            if self.trainable and infer_parameter:
+                raise ValueError("Gradients cannot be computed w.r.t. an inferred sigma value")
+            sigma = self.init_fn_sigma(x, y, dist)
+            tau = self.init_fn_tau(x, y, dist)
+            with torch.no_grad():
+                self.log_sigma.copy_(sigma.log().clone())
+                self.log_tau.copy_(tau.log().clone())
+            self.init_required = False
+
+        if len(self.sigma) > 1:
+            if len(self.sigma) == len(self.tau):
+                kernel_mat = []
+                for i in range(len(self.sigma)):
+                    kernel_mat.append(torch.exp(-2 * torch.square(
+                        torch.sin(torch.as_tensor(np.pi) * dist / self.tau[i])) / (self.sigma[i] ** 2)))
+                kernel_mat = torch.stack(kernel_mat, dim=0).mean(dim=0)
+            else:
+                raise ValueError("Length of sigma and alpha must be equal")
+        else:
+            kernel_mat = torch.exp(-2 * torch.square(
+                torch.sin(torch.as_tensor(np.pi) * dist / self.tau)) / (self.sigma ** 2))
+        return kernel_mat
+
+
+class LocalPeriodic(BaseKernel):
```

Comment:
Could this be defined as a product kernel of `Periodic` and `GaussianRBF`?

Reply:
Yes indeed, was writing it there just to make myself familiar with the kernel implementation logistic, will remove it later.